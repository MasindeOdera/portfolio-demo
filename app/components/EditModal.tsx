import { useState } from 'react';
import styled from 'styled-components';
import ModalContainer from './ModalContainer';
import { theme } from '../styles/theme';

type Project = {
    id: number;
    project: string;
    image: string;
    description: string;
};

interface EditModalProps {
  project: Project;
  onCancel: () => void;
  onSave: (updatedProject: Project) => void;
}

const FormContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 10px;
`;

const Label = styled.label`
  font-size: 0.9rem;
  font-weight: bold;
  color: ${theme.colors.primary};
  margin-bottom: 5px;
`;

const Input = styled.input`
  padding: 10px;
  border-radius: 5px;
  border: 1px solid ${theme.colors.detail};
  font-size: 1rem;
`;

const TextArea = styled.textarea`
  padding: 10px;
  border-radius: 5px;
  border: 1px solid ${theme.colors.detail};
  font-size: 1rem;
  height: 80px;
  resize: none;
`;

const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-between;
  margin-top: 20px;
`;

const Button = styled.button`
  width: 45%;
  padding: 10px 20px;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  font-weight: bold;

  &:nth-child(1) {
    background-color: ${theme.colors.success};
    color: ${theme.colors.text};
  }

  &:nth-child(2) {
    background-color: ${theme.colors.background};
    color: ${theme.colors.text};
  }

  &:hover {
    opacity: 0.9;
  }
`;

export default function EditModal({ project, onCancel, onSave }: EditModalProps) {
  const [name, setName] = useState(project.project);
  const [description, setDescription] = useState(project.description);

  return (
    <ModalContainer title="Edit Project">
      <FormContainer>
        <Label htmlFor="title">Title</Label>
        <Input 
          id="title" 
          value={name} 
          onChange={(e) => setName(e.target.value)} 
        />
        
        <Label htmlFor="description">Description</Label>
        <TextArea 
          id="description" 
          value={description} 
          onChange={(e) => setDescription(e.target.value)} 
        />
      </FormContainer>

      <ButtonContainer>
        <Button onClick={() => onSave({ ...project, project: name, description })}>
          Save
        </Button>
        <Button onClick={onCancel}>
          Cancel
        </Button>
      </ButtonContainer>
    </ModalContainer>
  );
}
