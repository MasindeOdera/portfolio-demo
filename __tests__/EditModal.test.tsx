import { render, screen, fireEvent } from '@testing-library/react';
import EditModal from '@/app/components/EditModal';

describe('EditModal Component', () => {
  const mockOnCancel = jest.fn();
  const mockOnSave = jest.fn();
  const mockProject = {
    id: 1,
    project: 'Test Project',
    image: 'https://example.com/image.jpg',
    description: 'This is a test project description.'
  };

  beforeEach(() => {
    render(
      <EditModal 
        project={mockProject} 
        onCancel={mockOnCancel} 
        onSave={mockOnSave} 
      />
    );
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <EditModal 
        project={mockProject} 
        onCancel={mockOnCancel} 
        onSave={mockOnSave} 
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders the correct initial values', () => {
    expect(screen.getByLabelText(/Title/i)).toHaveValue('Test Project');
    expect(screen.getByLabelText(/Description/i)).toHaveValue('This is a test project description.');
  });

  test('updates input values correctly', () => {
    const titleInput = screen.getByLabelText(/Title/i);
    const descriptionInput = screen.getByLabelText(/Description/i);

    fireEvent.change(titleInput, { target: { value: 'Updated Project' } });
    fireEvent.change(descriptionInput, { target: { value: 'Updated project description.' } });

    expect(titleInput).toHaveValue('Updated Project');
    expect(descriptionInput).toHaveValue('Updated project description.');
  });

  test('calls onCancel when Cancel button is clicked', () => {
    const cancelButton = screen.getByText(/Cancel/i);
    fireEvent.click(cancelButton);
    expect(mockOnCancel).toHaveBeenCalled();
  });

  test('calls onSave with correct data when Save button is clicked', () => {
    const saveButton = screen.getByText(/Save/i);
    const titleInput = screen.getByLabelText(/Title/i);
    const descriptionInput = screen.getByLabelText(/Description/i);

    fireEvent.change(titleInput, { target: { value: 'Updated Project' } });
    fireEvent.change(descriptionInput, { target: { value: 'Updated project description.' } });
    fireEvent.click(saveButton);

    expect(mockOnSave).toHaveBeenCalledWith({
      ...mockProject,
      project: 'Updated Project',
      description: 'Updated project description.'
    });
  });
});
