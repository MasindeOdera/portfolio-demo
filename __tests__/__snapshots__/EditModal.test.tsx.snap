// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditModal Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-blHHSb bkqoCW"
  >
    <h2
      class="sc-gtLWhw jidGLU"
    >
      Edit Project
    </h2>
    <div
      class="sc-egkSDF leRVCI"
    >
      <div
        class="sc-fAUdSK icfotn"
      >
        <label
          class="sc-dntaoT gRrdXS"
          for="title"
        >
          Title
        </label>
        <input
          class="sc-ivxoEo iDpIZj"
          id="title"
          value="Test Project"
        />
        <label
          class="sc-dntaoT gRrdXS"
          for="description"
        >
          Description
        </label>
        <textarea
          class="sc-ghWlax iwiYDP"
          id="description"
        >
          This is a test project description.
        </textarea>
      </div>
      <div
        class="sc-kLhKbu egrFSz"
      >
        <button
          class="sc-ixGGxD jYaIng"
        >
          Save
        </button>
        <button
          class="sc-ixGGxD jYaIng"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
